Add torch.library.register_autograd

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124299
* #124200
* #124199
* #124134
* #124180
* #124092
* #124089
* __->__ #124071
* #124066
* #124065
* #124064
* #123937

Allows registering autograd for all custom op entry points:
- the new-style custom op API (custom_op)
- the old-style torch.library APIs
- C++ operator registration

Test Plan:
- tests

zou3519 committed:
Add torch.library.register_autograd

Allows registering autograd for all custom op entry points:
- the new-style custom op API (@custom_op)
- the old-style torch.library APIs
- C++ operator registration

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.register_autograd"


Allows registering autograd for all custom op entry points:
- the new-style custom op API (custom_op)
- the old-style torch.library APIs
- C++ operator registration

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.register_autograd"


Allows registering autograd for all custom op entry points:
- the new-style custom op API (custom_op)
- the old-style torch.library APIs
- C++ operator registration

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.register_autograd"


Allows registering autograd for all custom op entry points:
- the new-style custom op API (custom_op)
- the old-style torch.library APIs
- C++ operator registration

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.register_autograd"


Allows registering autograd for all custom op entry points:
- the new-style custom op API (custom_op)
- the old-style torch.library APIs
- C++ operator registration

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.register_autograd"


Allows registering autograd for all custom op entry points:
- the new-style custom op API (custom_op)
- the old-style torch.library APIs
- C++ operator registration

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.register_autograd"


Allows registering autograd for all custom op entry points:
- the new-style custom op API (custom_op)
- the old-style torch.library APIs
- C++ operator registration

Test Plan:
- tests

[ghstack-poisoned]

albanD commented on torch/library.py:
```diff
@@ -553,6 +554,82 @@ def register(func):
+
+    In order for an operator to work with autograd, you need to register
+    a backward formula. There are two pieces to this:
+    1. You must tell us what we need to save from the forward pass for
```
Is setup_context actually a must? I would suggest
1. you must tell us how to compute gradients...
2. If you need any of the value from the forward there, you can use setup_ctx to ... 

WDYT?

zou3519 replied:
It's not a must, but most people need it. I'll update the wording to that, but I'm not sure if I should switch the order of the args to register_autograd (it's fine if the user passes None as the setup_context)

albanD commented on torch/library.py:
```diff
@@ -553,6 +554,82 @@ def register(func):
+    This is the "backward" function.
+
+    ``setup_context_fn(ctx, inputs, output)`` runs during the forward pass.
+    Please save quantities needed for backward onto the ``ctx`` object via
```
Not sure if I asked this in another PR but should we explicitly document the ctx here as being a custom Function ctx and link to that doc: https://pytorch.org/docs/stable/autograd.html#context-method-mixins ?

zou3519 commented on torch/_library/custom_ops.py:
```diff
@@ -391,7 +391,7 @@ def fake_impl(*args, **kwargs):
         lib._register_fake(self._name, fake_impl)
 
-        autograd_impl = _library.autograd.make_autograd_impl(self)
+        autograd_impl = _library.autograd.make_autograd_impl(self._opoverload, self)
```
Yes, via the power of duck typing

zou3519 replied:
I'll figure out how to update the typing (we should be able to type this via python protocols)

zou3519 committed:
Update on "Add torch.library.register_autograd"


Allows registering autograd for all custom op entry points:
- the new-style custom op API (custom_op)
- the old-style torch.library APIs
- C++ operator registration

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.register_autograd"


Allows registering autograd for all custom op entry points:
- the new-style custom op API (custom_op)
- the old-style torch.library APIs
- C++ operator registration

Test Plan:
- tests

[ghstack-poisoned]

zou3519 commented:
Improved the docs and updated the typing

zou3519 committed:
Update on "Add torch.library.register_autograd"


Allows registering autograd for all custom op entry points:
- the new-style custom op API (custom_op)
- the old-style torch.library APIs
- C++ operator registration

Test Plan:
- tests

[ghstack-poisoned]

albanD commented on torch/_library/autograd.py:
```diff
@@ -1,26 +1,39 @@
+    _backward_fn: Optional[Callable]
+
+
+# op actually has type "torch._ops.OpOverload", but we've got some circular import issues.
+def make_autograd_impl(op: Any, info: InfoProtocol) -> Callable:
```
```suggestion
def make_autograd_impl(op: "torch._ops.OpOverload", info: InfoProtocol) -> Callable:
```

That's why strings as types are allowed!

zou3519 replied:
mypy is smarter nowadays; it will complain that torch isn't imported if we do "torch._ops.OpOverload" :/

albanD replied:
Ho and it will complain that torch is not used as well?
We can have an `import torch` at the top of the file right?

zou3519 replied:
I got away with `from . import _ops`, but `import torch` at the top could trigger circular import issues

albanD reviewed (approved):
nit on typing

zou3519 committed:
Update on "Add torch.library.register_autograd"


Allows registering autograd for all custom op entry points:
- the new-style custom op API (custom_op)
- the old-style torch.library APIs
- C++ operator registration

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.register_autograd"


Allows registering autograd for all custom op entry points:
- the new-style custom op API (custom_op)
- the old-style torch.library APIs
- C++ operator registration

Test Plan:
- tests

[ghstack-poisoned]
